### PR TITLE
Feature/add request decorator

### LINKED ISSUE #12: Refactor to provide methods rather than register hook
Currently we are registering an `onRequest` hook when the plugin is registered. We should instead register methods (as decorators) to provide more flexibility to the user about which routes they wish to verify.

Using https://github.com/fastify/fastify-jwt as an example, I think we should:

- register `verificationHook` as a request decorator
- provide an example of using it as a route level `preValidation` hook (https://www.fastify.io/docs/latest/Hooks/#route-level-hooks)

### PR DESCRIPTION
**Pull request by jeff-sexton**
resolves #12 
### Summary
- Refactored Plugin registration function to add a request decorator instead of adding a hook
- Updated base example to utilize the plugin as a `preValidation` hook.
- Added new example to demonstrate using the plugin request decorator in a route level `preValidation` hook.
- Updated Readme to show the different use cases.

### Test Plan
- run global hook example `npm run example:hook -- -l info -w`
  - Verify that requests with valid Signature headers are appropriately validated by the plugin.
  - Change request header and confirm that the plugin returns the verification error message.
  - Example HTTPie request:
 ```sh
http POST localhost:3000 Signature:'keyId="key-a", headers="(request-target) (created) (expires) host", signature="Nstb3q49bne90WmkbiZ9eyRtKCUmOWvc/kFyw/ftfLtqAXfRZGBk+hSJkXs0GRRm4Q/58c/pdIKj5DND11/fwQ==", created=1402170695, expires=1402170895'
```
- run route level decorator example `npm run example:decorator -- -l info -w`
  - Verify that any request sent to `/` is passed and the fastify server responds with `{ hello: 'no validation needed'}`.
  - Verify that requests with valid Signature headers sent to /foo are validated by the plugin. Fastify server will respond with `{ hello: 'validated HMAC' }`.
  - Change request header sent to `/foo` and confirm that the plugin returns the verification error message.
  - Example HTTPie request:
```sh
http POST localhost:3000/foo Signature:'keyId="key-a", headers="(request-target) (created) (expires) host", signature="fKj7/7DbxLi1RkMqBrPB6HJ6dXUThRM4nwSErMo3ni1SuXZjTEpNDiZozUMUYqpcGssakLQl2y8ryBz0U5oxlg==", created=1402170695, expires=1402170895'
```
- **NOTE:** Any implementation currently using the `fastify-hmac` plugin will need to be updated to use this new request decorator pattern.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -3,9 +3,12 @@
 const fp = require('fastify-plugin')
 const { plugin } = require('./lib')
 
-module.exports = fp(function (fastify, options, next) {
-  fastify.addHook('onRequest', plugin(options))
+function fastifyHMAC (fastify, options, next) {
+  fastify.decorateRequest('HMACValidate', plugin(options))
+
```
Comment (togmund) [inline index.js:8]: Really my only quibble is the naming of this decorator. Function names _generally_ contain verbs and describe the action that will take place. `HMACValidate` doesn't read to me as an "action".

It looks as though you had this same instinct, as you've named it the way I would have expected it to be named when you decorated the example: [validateHMAC](https://github.com/autotelic/fastify-hmac/pull/17/files#diff-ea03bd13687ba7145ba463540ccb8216R12).

I might use the `validateHMAC` in the plugin then decorate your example with the same name, or another similar action-name like `verifyHMAC`.

**Comment by togmund** [comment]
This is really good work!

The code is concise and readable, the example is easy to grok, plus the test plan was a breeze and responded exactly as I would have expected.

My only quibble about the decorator name is an opinion, so feel free to get more feedback on that if you want a broader set of opinions.

LGTM 😎

File: index.js (hunk 1 above)
Comment (jeff-sexton) [reply index.js]: This is a good point. I've updated the naming accordingly. Thanks!

File: README.md [hunk 2]
```diff
@@ -12,49 +16,146 @@ npm i @autotelic/fastify-hmac
...
-  - A `sharedSecret` string
-  - A `getAlgorithm` function that returns an algorithm string
-  - A `getDigest` function that returns a digest string
-
-### Example:
+- Register plugin. This will decorate your `fastify` instance with a request method `HMACValidate`.
+  - During registration, provide a configuration object that contains the following:
+    - A `sharedSecret` string
+    - A `getAlgorithm` function that returns an algorithm string
+    - A `getDigest` function that returns a digest string
```
Comment (jkirkpatrick24) [inline README.md:23]: Lets update these options now that getDigest has been implemented. 👍

Comment (jeff-sexton) [reply README.md]: I updated the naming and added details for `getDigest` and `digestEncoding`.

**Pull request merged**